kernel/aster-nix/src/process/program_loader/elf/init_stack.rs

```diff
@@ -93,9 +99,13 @@ impl InitStack {
         }
     }
 
-    /// This function only work for first process
     pub fn new_default_config(argv: Vec<CString>, envp: Vec<CString>) -> Self {
-        let init_stack_top = INIT_STACK_BASE - PAGE_SIZE;
+        let nr_pages_padding = {
+            let mut random_nr_pages_padding: u8 = 0;
+            getrandom::getrandom(random_nr_pages_padding.as_bytes_mut()).unwrap();
+            random_nr_pages_padding as usize + 1
```

Comment:
Adding randomization without adding ways to disable it can make debugging more difficult. So I would argue against adding such randomization for the kernel offset and code/data segments.

But since I can't think of any cases where the stack location matters (who knows? maybe someday it does), I'm neutral with this change, but I still think it's a point worth mentioning.

Reply:
I once thought we could add a UASLR feature gate but the second thought yields it unnecessary:
 - If a user space program want to know the stack base to allow debugging, just info mappings or read procfs; they shall not rebuild the kernel;
 - If we want to debug it we can just read it.

However, a KASLR feature gate is necessary. I would think adding it in the future for easier debugging.

Reply:
By default, GDB disables ASLR using the [personality](https://man7.org/linux/man-pages/man2/personality.2.html) system call with the `ADDR_NO_RANDOMIZE` argument. In addition, we can globally disable ASLR in Linux using `/proc/sys/kernel/randomize_va_space` in procfs or `norandmaps` on the kernel command line. As for kernel randomization, there is a `nokaslr` option on the Linux kernel command line.

> * If a user space program want to know the stack base to allow debugging, just info mappings or read procfs; they shall not rebuild the kernel;

It's a tedious process because we have to do "info mappings or read procfs" every time we (re)run the program we want to debug, which can be dozens of times before the bug is finally caught.

> * If we want to debug it we can just read it.

I don't know if there is a mature way to debug user space programs in Asterinas yet. (I don't think we support GDB yet, do we?) So I used to do this by making my user space program the `init` process, and then setting breakpoints directly at the user space addresses using the GDB attached to QEMU. The problem is that when the user space segments are always randomly placed, I will have a hard time figuring out what the address is and then changing the previous breakpoints accordingly.

Making _only_ the stack location random might be a fair change, and its impact is minimal compared to randomizing the code/data segments. So perhaps this discussion is somehow outside the scope of this PR? I'm not sure.

Reply:
Yep, in this PR only the stack is randomized and a whole system for configuring the ASLR feature (just impacting the stack) seems an overkill. And currently the user program's code and data segment are placed just as where the ELF says (except for the shared library to which we cannot do so).

Your idea is valuable and I do accept that we need kernel command line/procfs configurations if actual ASLR is fully introduced in the future.

Reply:
Yes, it does not seem right to do that in this PR, of course. That is also not what I mean.